File: src/Eleicao.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;

public class Eleicao {

  private LocalDate dataEleicao;
  private int codigoCidade;
  private int quantidadeEleitos;
  private Map<String, Partido> partidos = new HashMap<>(); // numero partido -> Partido
  private Map<String, Candidato> candidatos = new HashMap<>(); // numero candidato -> Candidato
  private DateTimeFormatter formatterDate = DateTimeFormatter.ofPattern("dd/MM/yyyy");

  public Eleicao(int codigoCidade, String dataEleicao) {
    this.codigoCidade = codigoCidade;
    this.dataEleicao = LocalDate.parse(dataEleicao, formatterDate);
  }

  public void processaVotacao(List<String[]> linhas, Map<String, Integer> indiceCabecalho) {
    for (String[] valores : linhas) {
      try {
        int codigoCidadeArquivo = Integer.parseInt(valores[indiceCabecalho.get("CD_MUNICIPIO")]);
        int codigoCargo = Integer.parseInt(valores[indiceCabecalho.get("CD_CARGO")]);

        if (codigoCidadeArquivo != this.codigoCidade || codigoCargo != 13)
          continue;

        int quantidadeVotos = Integer.parseInt(valores[indiceCabecalho.get("QT_VOTOS")]);
        int numeroCandidato = Integer.parseInt(valores[indiceCabecalho.get("NR_VOTAVEL")]);

        if (numeroCandidato >= 10000 && numeroCandidato <= 99999) {
          Candidato candidato = candidatos.get(String.valueOf(numeroCandidato));
          if (candidato != null) {
            candidato.incrementaQuantidadeVotos(quantidadeVotos);
            Partido partido = candidato.getPartido();
            if (partido != null) {
              partido.incrementaVotosNominais(quantidadeVotos);
            }
          }
        } else if (numeroCandidato >= 10 && numeroCandidato <= 99
            && !(numeroCandidato >= 95 && numeroCandidato <= 98)) {
          String partidoKey = String.valueOf(numeroCandidato);
          if (partidos.containsKey(partidoKey)) {
            Partido partido = partidos.get(partidoKey);
            partido.incrementaVotosLegenda(quantidadeVotos);
          }
        }
      } catch (Exception e) {
        System.out.println("Erro ao processar linha de votação: " + Arrays.toString(valores));
        e.printStackTrace();
      }
    }
  }

  public void processaCandidatosPartidos(List<String[]> linhas, Map<String, Integer> indiceCabecalho) {
    for (String[] valores : linhas) {
      try {
        if (!ehCandidatoValido(valores, indiceCabecalho))
          continue;

        Partido partido = getOrCreatePartido(valores, indiceCabecalho);
        Candidato candidato = criaCandidato(valores, partido, indiceCabecalho);

        partido.insereCandidato(candidato);
        this.candidatos.put(candidato.getNumeroCandidato(), candidato);

        if (candidato.getEleito()) {
          this.quantidadeEleitos++;
        }
      } catch (Exception e) {
        System.out.println("Erro ao processar linha de candidato: " + Arrays.toString(valores));
        e.printStackTrace();
      }
    }
  }

  private boolean ehCandidatoValido(String[] valores, Map<String, Integer> indiceCabecalho) {
    int codigoCidadeArquivo = Integer.parseInt(valores[indiceCabecalho.get("SG_UE")]);
    int codigoCargo = Integer.parseInt(valores[indiceCabecalho.get("CD_CARGO")]);
    return codigoCidadeArquivo == this.codigoCidade && codigoCargo == 13;
  }

  private Partido getOrCreatePartido(String[] valores, Map<String, Integer> indiceCabecalho) {
    String numeroPartido = valores[indiceCabecalho.get("NR_PARTIDO")];
    String siglaPartido = valores[indiceCabecalho.get("SG_PARTIDO")];

    if (!partidos.containsKey(numeroPartido)) {
      partidos.put(numeroPartido, new Partido(numeroPartido, siglaPartido));
    }
    return partidos.get(numeroPartido);
  }

  private Candidato criaCandidato(String[] valores, Partido partido, Map<String, Integer> indiceCabecalho) {
    String numeroCandidato = valores[indiceCabecalho.get("NR_CANDIDATO")];
    String nomeCandidatoUrna = valores[indiceCabecalho.get("NM_URNA_CANDIDATO")];
    String dataNascimento = valores[indiceCabecalho.get("DT_NASCIMENTO")];

    int numFederacao = Integer.parseInt(valores[indiceCabecalho.get("NR_FEDERACAO")]);
    int genero = Integer.parseInt(valores[indiceCabecalho.get("CD_GENERO")]);
    int codigoCandidatoEleito = Integer.parseInt(valores[indiceCabecalho.get("CD_SIT_TOT_TURNO")]);

    boolean candidatoEleito = codigoCandidatoEleito == 2 || codigoCandidatoEleito == 3;

    Candidato candidato = new Candidato();
    candidato.setNumeroCandidato(numeroCandidato);
    candidato.setNomeUrna(nomeCandidatoUrna);
    candidato.setPartido(partido);
    candidato.setParticipaFederacao(numFederacao != -1);
    candidato.setDataNascimento(LocalDate.parse(dataNascimento, formatterDate));
    candidato.setEleito(candidatoEleito);

    if (genero == 2) {
      candidato.setGenero(Genero.MASCULINO);
    } else if (genero == 4) {
      candidato.setGenero(Genero.FEMININO);
    }

    return candidato;
  }

  public Collection<Candidato> getCandidatos() {
    return candidatos.values();
  }

  public Collection<Partido> getPartidos() {
    return partidos.values();
  }

  public int getQuantidadeEleitos() {
    return quantidadeEleitos;
  }

  public LocalDate getDataEleicao() {
    return dataEleicao;
  }
}

File: implementacao-c++/eleicao.cpp
#include "eleicao.hpp"

#include <algorithm>
#include <fstream>
#include <sstream>
#include <vector>

#include "iso8859_1_to_utf8.hpp"

Eleicao::Eleicao(int codigoCidade, const string& arquivoCandidatos, const string& arquivoVotacao, const string& dataEleicao) {
  this->codigoCidade = codigoCidade;
  this->arquivoCandidatos = arquivoCandidatos;
  this->arquivoVotacao = arquivoVotacao;
  // Converte a string de data para o formato tm
  strptime(dataEleicao.c_str(), "%d/%m/%Y", &this->dataEleicao);
}

void Eleicao::processaCandidatos() {
  ifstream arquivo(this->arquivoCandidatos, ios::in);
  if (!arquivo.is_open()) {
    cerr << "Erro ao abrir o arquivo de candidatos" << endl;
    return;
  }
  // Cria um map de índices das colunas
  map<string, int> indiceCabecalhos = {};

  string linha;
  bool isCabecalho = true;
  while (getline(arquivo, linha)) {
    // Converte a linha de ISO-8859-1 para UTF-8 e remove as aspas
    string linhaUtf8 = iso_8859_1_to_utf8(linha);
    linhaUtf8.erase(remove(linhaUtf8.begin(), linhaUtf8.end(), '\"'), linhaUtf8.end());

    // Transforma em um vetor de strings
    vector<string> valores;
    stringstream ss(linhaUtf8);
    string campo;
    while (getline(ss, campo, ';'))
      valores.push_back(campo);

    if (isCabecalho) {
      // Preenche o map de índices das colunas (size_t para remover warning)
      for (size_t i = 0; i < valores.size(); i++)
        indiceCabecalhos[valores[i]] = i;
      isCabecalho = false;
      continue;
    }

    // Obtem informacoes do partido
    string numeroPartido = valores[indiceCabecalhos["NR_PARTIDO"]];
    string nomePartido = valores[indiceCabecalhos["SG_PARTIDO"]];

    Partido* p = nullptr;

    // Verifica se o partido ja foi inserido no map
    if (this->partidos.find(numeroPartido) == this->partidos.end()) {
      p = new Partido;
      p->setNumero(numeroPartido);
      p->setSigla(nomePartido);
      this->partidos[numeroPartido] = p;
    } else {
      p = this->partidos[numeroPartido];
    }

    int codigoCidadeArquivo = stoi(valores[indiceCabecalhos["SG_UE"]]);
    int codigoCargo = stoi(valores[indiceCabecalhos["CD_CARGO"]]);
    if (codigoCidadeArquivo != this->codigoCidade || codigoCargo != 13)
      continue;

    // Informacoes do candidato
    int numFederacao = stoi(valores[indiceCabecalhos["NR_FEDERACAO"]]);
    int genero = stoi(valores[indiceCabecalhos["CD_GENERO"]]);
    string numeroCandidato = valores[indiceCabecalhos["NR_CANDIDATO"]];
    string nomeCandidatoUrna = valores[indiceCabecalhos["NM_URNA_CANDIDATO"]];
    string dataNascimento = valores[indiceCabecalhos["DT_NASCIMENTO"]];
    bool candidatoEleito = false;

    int codigoCandidatoEleito = stoi(valores[indiceCabecalhos["CD_SIT_TOT_TURNO"]]);
    if (codigoCandidatoEleito == -1) {
      continue;
    } else if (codigoCandidatoEleito == 2 || codigoCandidatoEleito == 3) {
      candidatoEleito = true;
      this->quantidadeEleitos++;
    }

    Candidato* candidato = new Candidato;

    candidato->setEleito(candidatoEleito);
    candidato->setNumeroCandidato(numeroCandidato);
    candidato->setNomeUrna(nomeCandidatoUrna);
    candidato->setPartido(*p);
    candidato->setParticipaFederacao(numFederacao != -1);
    candidato->setDataNascimento(dataNascimento);

    if (genero == 2) {
      candidato->setGenero(Genero::MASCULINO);
    } else if (genero == 4) {
      candidato->setGenero(Genero::FEMININO);
    }

    this->candidatos[numeroCandidato] = candidato;
    p->insereCandidato(*candidato);
  }

  arquivo.close();
}

void Eleicao::processaVotacao() {
  ifstream arquivo(this->arquivoVotacao, ios::in);
  if (!arquivo.is_open()) {
    cerr << "Erro ao abrir o arquivo de candidatos" << endl;
    return;
  }
  // Cria um map de índices das colunas
  map<string, int> indiceCabecalhos = {};

  string linha;
  bool isCabecalho = true;
  while (getline(arquivo, linha)) {
    // Converte a linha de ISO-8859-1 para UTF-8 e remove as aspas
    string linhaUtf8 = iso_8859_1_to_utf8(linha);
    linhaUtf8.erase(remove(linhaUtf8.begin(), linhaUtf8.end(), '\"'), linhaUtf8.end());

    // Transforma em um vetor de strings
    vector<string> valores;
    stringstream ss(linhaUtf8);
    string campo;
    while (getline(ss, campo, ';'))
      valores.push_back(campo);

    if (isCabecalho) {
      // Preenche o map de índices das colunas (size_t para remover warning)
      for (size_t i = 0; i < valores.size(); i++)
        indiceCabecalhos[valores[i]] = i;
      isCabecalho = false;
      continue;
    }

    int codigoCidadeArquivo = stoi(valores[indiceCabecalhos["CD_MUNICIPIO"]]);
    int codigoCargo = stoi(valores[indiceCabecalhos["CD_CARGO"]]);
    if (codigoCidadeArquivo != this->codigoCidade || codigoCargo != 13)
      continue;

    int qtdVotos = stoi(valores[indiceCabecalhos["QT_VOTOS"]]);
    int numCandidato = stoi(valores[indiceCabecalhos["NR_VOTAVEL"]]);

    if (numCandidato >= 10000 && numCandidato <= 99999) {
      string numeroCandidato = to_string(numCandidato);
      if (this->candidatos.find(numeroCandidato) != this->candidatos.end()) {
        Candidato* candidato = this->candidatos[numeroCandidato];
        candidato->incrementaVotos(qtdVotos);
        Partido* p = candidato->getPartido();
        p->incrementaVotosNominais(qtdVotos);
      }
    } else if (numCandidato >= 10 && numCandidato <= 99 && !(numCandidato >= 95 && numCandidato <= 98)) {
      // Nao eh candidato, eh partido
      if (this->partidos.find(to_string(numCandidato)) != this->partidos.end()) {
        Partido* p = this->partidos[to_string(numCandidato)];
        p->incrementaVotosLegenda(qtdVotos);
      }
    }
  }

  arquivo.close();
}

const map<string, Partido*>& Eleicao::getPartidos() const {
  return this->partidos;
}

const map<string, Candidato*>& Eleicao::getCandidatos() const {
  return this->candidatos;
}

int Eleicao::getQuantidadeEleitos() const {
  return this->quantidadeEleitos;
}

tm Eleicao::getDataEleicao() const {
  return this->dataEleicao;
}

// Libera memoria alocada para os partidos e candidatos
Eleicao::~Eleicao() {
  for (auto& [_, partido] : this->partidos)
    delete partido;
  for (auto& [_, candidato] : this->candidatos)
    delete candidato;
}
